File: src/backend/decoders/sip.py
def decodeSIP(rawPayload):
    """
    Decode SIP message fields from raw payload bytes.
    Parses the first line and common headers (From, To, Call-ID, Authorization).
    Returns a dict with both display-friendly keys and dot-notation keys for message
    type, method/status, and headers, or None if the payload is not a SIP message or
    decoding fails.
    """
    sipMethods = {
        "INVITE",
        "ACK",
        "BYE",
        "CANCEL",
        "REGISTER",
        "OPTIONS",
        "SUBSCRIBE",
        "NOTIFY",
        "REFER",
        "INFO",
        "UPDATE",
        "PRACK",
    }
    try:
        text = rawPayload.decode(errors="ignore")
        lines = text.split("\r\n") if "\r\n" in text else text.split("\n")
        if not lines:
            return None
        firstLine = lines[0].strip()
        isSipResponse = firstLine.startswith("SIP/")
        isSipRequest = (
            firstLine.split(" ")[0] in sipMethods if " " in firstLine else False
        )
        if not isSipResponse and not isSipRequest:
            return None
        headers = {}
        for line in lines[1:]:
            if ": " in line:
                key, _, val = line.partition(": ")
                headers[key.strip()] = val.strip()

        authorization = headers.get("Authorization", "")
        proxyAuthorization = headers.get("Proxy-Authorization", "")

        if isSipRequest:
            parts = firstLine.split(" ", 2)
            method = parts[0]
            requestUri = parts[1] if len(parts) > 1 else "Unknown"
            result = {
                "Type": "Request",
                "sip.type": "Request",
                "Method": method,
                "sip.method": method,
                "Request URI": requestUri,
                "sip.uri": requestUri,
                "From": headers.get("From", "Unknown"),
                "sip.from": headers.get("From", "Unknown"),
                "To": headers.get("To", "Unknown"),
                "sip.to": headers.get("To", "Unknown"),
                "Call-ID": headers.get("Call-ID", "Unknown"),
                "sip.call_id": headers.get("Call-ID", "Unknown"),
            }
            if authorization:
                result["Authorization"] = authorization
                result["sip.authorization"] = authorization
            if proxyAuthorization:
                result["Proxy-Authorization"] = proxyAuthorization
                result["sip.proxy_authorization"] = proxyAuthorization
            return result
        else:
            parts = firstLine.split(" ", 2)
            statusCode = parts[1] if len(parts) > 1 else "Unknown"
            statusMsg = parts[2] if len(parts) > 2 else "Unknown"
            result = {
                "Type": "Response",
                "sip.type": "Response",
                "Status Code": statusCode,
                "sip.status_code": statusCode,
                "Status Message": statusMsg,
                "sip.status_msg": statusMsg,
                "From": headers.get("From", "Unknown"),
                "sip.from": headers.get("From", "Unknown"),
                "To": headers.get("To", "Unknown"),
                "sip.to": headers.get("To", "Unknown"),
                "Call-ID": headers.get("Call-ID", "Unknown"),
                "sip.call_id": headers.get("Call-ID", "Unknown"),
            }
            if authorization:
                result["Authorization"] = authorization
                result["sip.authorization"] = authorization
            if proxyAuthorization:
                result["Proxy-Authorization"] = proxyAuthorization
                result["sip.proxy_authorization"] = proxyAuthorization
            return result
    except Exception:
        return None

File: src/ui/decoders/conv/sip.js
// SIP Conv decoder: parses the first line of a SIP message (request or
// response) and a curated set of common headers, with support for compact
// header forms ("f:" = From, "t:" = To, etc.).

const SIP_METHODS = new Set([
    "INVITE",
    "ACK",
    "BYE",
    "CANCEL",
    "REGISTER",
    "OPTIONS",
    "SUBSCRIBE",
    "NOTIFY",
    "REFER",
    "INFO",
    "UPDATE",
    "PRACK",
    "MESSAGE",
    "PUBLISH",
]);

const COMPACT_HEADER_NAMES = {
    f: "from",
    t: "to",
    i: "call-id",
    m: "contact",
    v: "via",
    l: "content-length",
    c: "content-type",
    r: "refer-to",
};

const INTERESTING_HEADERS = [
    ["from", "From"],
    ["to", "To"],
    ["call-id", "Call-ID"],
    ["cseq", "CSeq"],
    ["via", "Via"],
    ["contact", "Contact"],
    ["max-forwards", "Max-Forwards"],
    ["user-agent", "User-Agent"],
    ["authorization", "Authorization"],
    ["proxy-authorization", "Proxy-Authorization"],
    ["route", "Route"],
    ["record-route", "Record-Route"],
    ["content-type", "Content-Type"],
    ["content-length", "Content-Length"],
    ["expires", "Expires"],
];

const REQUEST_LINE_RE = /^([A-Z]+)\s+(\S+)\s+SIP\/([\d.]+)$/i;
const RESPONSE_LINE_RE = /^SIP\/([\d.]+)\s+(\d{3})(?:\s+(.*))?$/i;
const BODY_PREVIEW_LIMIT = 220;
const HEADER_VALUE_LIMIT = 180;

function truncateField(value, limit) {
    if (typeof value !== "string") return "";
    const trimmed = value.trim();
    if (!trimmed) return "";
    return trimmed.length > limit ? `${trimmed.slice(0, limit)}...` : trimmed;
}

function decodeSipFromBytes(bytes) {
    if (!(bytes instanceof Uint8Array) || bytes.length === 0) return null;

    const text = new TextDecoder("utf-8", { fatal: false }).decode(bytes);
    const lines = text.split(/\r?\n/);
    if (!lines.length) return null;

    const firstLine = (lines[0] || "").trim();
    if (!firstLine) return null;

    const requestMatch = firstLine.match(REQUEST_LINE_RE);
    const responseMatch = firstLine.match(RESPONSE_LINE_RE);
    const isRequest = Boolean(requestMatch && SIP_METHODS.has(requestMatch[1].toUpperCase()));
    const isResponse = Boolean(responseMatch);
    if (!isRequest && !isResponse) return null;

    const headerLines = [];
    let bodyStartIndex = lines.length;
    for (let i = 1; i < lines.length; i += 1) {
        const rawLine = lines[i] || "";
        if (!rawLine.trim()) {
            bodyStartIndex = i + 1;
            break;
        }
        if (/^[ \t]/.test(rawLine) && headerLines.length) {
            headerLines[headerLines.length - 1] += ` ${rawLine.trim()}`;
            continue;
        }
        headerLines.push(rawLine);
    }

    const headerMap = new Map();
    headerLines.forEach((line) => {
        const separator = line.indexOf(":");
        if (separator <= 0) return;
        const rawName = line.slice(0, separator).trim();
        const value = line.slice(separator + 1).trim();
        if (!rawName || !value) return;
        const lowered = rawName.toLowerCase();
        const normalizedName = COMPACT_HEADER_NAMES[lowered] || lowered;
        if (!headerMap.has(normalizedName)) headerMap.set(normalizedName, []);
        headerMap.get(normalizedName).push(value);
    });

    const getHeaderValue = (name) => {
        const values = headerMap.get(String(name || "").toLowerCase());
        if (!Array.isArray(values) || !values.length) return "";
        return values.join(" | ");
    };

    const fields = [];
    if (isRequest && requestMatch) {
        fields.push(
            { name: "Type", value: "Request" },
            { name: "Method", value: requestMatch[1].toUpperCase() },
            { name: "Request URI", value: requestMatch[2] || "N/A" },
            { name: "SIP Version", value: requestMatch[3] || "N/A" },
        );
    }
    if (isResponse && responseMatch) {
        fields.push(
            { name: "Type", value: "Response" },
            { name: "SIP Version", value: responseMatch[1] || "N/A" },
            { name: "Status Code", value: responseMatch[2] || "N/A" },
            { name: "Reason Phrase", value: responseMatch[3] || "N/A" },
        );
    }

    INTERESTING_HEADERS.forEach(([headerKey, label]) => {
        const value = truncateField(getHeaderValue(headerKey), HEADER_VALUE_LIMIT);
        if (value) fields.push({ name: label, value });
    });

    const bodyText = lines.slice(bodyStartIndex).join("\n").trim();
    if (bodyText) {
        fields.push({
            name: "Body Preview",
            value: truncateField(bodyText, BODY_PREVIEW_LIMIT),
        });
    }

    return fields.length ? { protocol: "SIP", fields } : null;
}

module.exports = { decodeSipFromBytes };
